File: rb/test/spacecraft_entity_test.rb
# Spacecraft entity test

require "minitest/autorun"
require "json"
require_relative "../StarTrek_sdk"
require_relative "runner"

class SpacecraftEntityTest < Minitest::Test
  def test_create_instance
    testsdk = StarTrekSDK.test(nil, nil)
    ent = testsdk.Spacecraft(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = spacecraft_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "spacecraft." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set STARTREK_TEST_SPACECRAFT_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    spacecraft_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.spacecraft")))
    spacecraft_ref01_data = nil
    if spacecraft_ref01_data_raw.length > 0
      spacecraft_ref01_data = Helpers.to_map(spacecraft_ref01_data_raw[0][1])
    end

    # LIST
    spacecraft_ref01_ent = client.Spacecraft(nil)
    spacecraft_ref01_match = {}

    spacecraft_ref01_list_result, err = spacecraft_ref01_ent.list(spacecraft_ref01_match, nil)
    assert_nil err
    assert spacecraft_ref01_list_result.is_a?(Array)

  end
end

def spacecraft_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "spacecraft", "SpacecraftTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = StarTrekSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["spacecraft01", "spacecraft02", "spacecraft03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["STARTREK_TEST_SPACECRAFT_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "STARTREK_TEST_SPACECRAFT_ENTID" => idmap,
    "STARTREK_TEST_LIVE" => "FALSE",
    "STARTREK_TEST_EXPLAIN" => "FALSE",
    "STARTREK_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["STARTREK_TEST_SPACECRAFT_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["STARTREK_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["STARTREK_APIKEY"],
      },
      extra || {},
    ])
    client = StarTrekSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["STARTREK_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["STARTREK_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: php/test/SpacecraftEntityTest.php
<?php
declare(strict_types=1);

// Spacecraft entity test

require_once __DIR__ . '/../startrek_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class SpacecraftEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = StarTrekSDK::test(null, null);
        $ent = $testsdk->Spacecraft(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = spacecraft_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "spacecraft." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set STARTREK_TEST_SPACECRAFT_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $spacecraft_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.spacecraft")));
        $spacecraft_ref01_data = null;
        if (count($spacecraft_ref01_data_raw) > 0) {
            $spacecraft_ref01_data = Helpers::to_map($spacecraft_ref01_data_raw[0][1]);
        }

        // LIST
        $spacecraft_ref01_ent = $client->Spacecraft(null);
        $spacecraft_ref01_match = [];

        $spacecraft_ref01_list_result = $spacecraft_ref01_ent->list($spacecraft_ref01_match, null);
        $this->assertIsArray($spacecraft_ref01_list_result);

    }
}

function spacecraft_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/spacecraft/SpacecraftTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = StarTrekSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["spacecraft01", "spacecraft02", "spacecraft03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("STARTREK_TEST_SPACECRAFT_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "STARTREK_TEST_SPACECRAFT_ENTID" => $idmap,
        "STARTREK_TEST_LIVE" => "FALSE",
        "STARTREK_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["STARTREK_TEST_SPACECRAFT_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["STARTREK_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new StarTrekSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["STARTREK_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["STARTREK_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
